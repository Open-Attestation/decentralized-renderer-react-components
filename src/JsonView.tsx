import React from "react";
import { EmotionJSX } from "@emotion/react/types/jsx-namespace";

const INDENT_PX = 40;
const MAX_STACK_COUNT = 10;
const MAX_CHARACTERS = 200;
const MAX_CHARACTERS_REVEAL_FIRST_N = 100;
const MAX_CHARACTERS_REVEAL_LAST_N = 100;

function JsonLine(key: string, val: any) {
  return (
    <div key={key}>
      <span style={{ color: "black", fontWeight: 700 }}>{key}: </span>
      <span style={{ color: "#057A55" }}>{val}</span>
    </div>
  );
}

function formatJson(variable: any, stackCount = 0, indent = 0, key = ""): EmotionJSX.Element | string {
  if (stackCount > MAX_STACK_COUNT) {
    /* nested 5 levels. stop showing */
    return "...";
  }

  if (typeof variable === "object" && variable !== null && !Array.isArray(variable)) {
    const jsonLines = [];
    for (const key of Object.keys(variable)) {
      if (Array.isArray(variable[key])) {
        /* Array items will be rendered differently, and alongside sibling attributes.
        KEY_TO_ARRAY_ITEM[0]: ...
        KEY_TO_ARRAY_ITEM[1]: ...
        SIBLING_ATTRIBUTE: ...

        indent = 0 to keep alignment with sibling attributes
        */
        jsonLines.push(formatJson(variable[key], stackCount + 1, 0, key));
      } else {
        /* if attribute is a nested object, display the key and indent.
        
        KEY_TO_NESTED_OBJ:
            ATTRIBUTE1: ...
            ATTRIBUTE2: ...
        */
        jsonLines.push(JsonLine(key, formatJson(variable[key], stackCount + 1, INDENT_PX)));
      }
    }
    return <div style={{ marginLeft: `${indent}px` }}>{jsonLines}</div>; // wrapping in div basically starts a new line
  }

  if (Array.isArray(variable)) {
    return (
      <div style={{ marginLeft: `${indent}px` }}>
        {/* if an attribute is an array, display elements as <KEY_NAME>[0]: <ATTRIBUTE_VAL> */}
        {variable.map((element, index) => JsonLine(`${key}[${index}]`, formatJson(element, stackCount + 1, INDENT_PX)))}
      </div> // wrapping in div basically starts a new line
    );
  }

  if (typeof variable === "string" && variable.length >= MAX_CHARACTERS) {
    return (
      variable.substring(0, MAX_CHARACTERS_REVEAL_FIRST_N) +
      "[...]" +
      variable.substring(variable.length - MAX_CHARACTERS_REVEAL_LAST_N)
    );
  }

  return String(variable);
}

export const JsonView: React.FunctionComponent<{ src: any }> = (props) => {
  return (
    <pre
      data-testid={"json-view"}
      style={{
        backgroundColor: "white",
        fontSize: "14px",
        lineHeight: "21px",
        whiteSpace: "pre-wrap",
        wordWrap: "break-word",
      }}
    >
      {formatJson(props.src)}
    </pre>
  );
};
