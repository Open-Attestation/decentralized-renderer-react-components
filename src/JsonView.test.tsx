import React from "react";
import { JsonView } from "./JsonView";
import { render } from "@testing-library/react";

describe("jsonView component", () => {
  it("should display json view of a json object with nested elements", () => {
    const screen = render(
      <JsonView
        src={{
          string: "test",
          number: 1,
          boolean: false,
          person: { name: "john doe", age: 35, isMarried: false },
          nestedObj: { anotherLayerOfNesting: { person2: { name: "jane doe", age: 20, isMarried: true } } },
          arrayOfObj: [
            { three: "three", four: "four" },
            { five: "five", six: "six" },
          ],
          arrayOfArr: [
            ["test3", "test4"],
            ["test5", "test6"],
          ],
          aVeryLongString:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
          aVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongKeyName:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
          aVeryNestedObj: {
            nest1: {
              nest2: { nest3: { nest4: { nest5: { nest6: { nest7: { nest8: { nest9: { nest10: "test" } } } } } } } },
            },
          },
        }}
      />
    );

    expect(screen.getByTestId("json-view")).toMatchSnapshot();
  });
});
