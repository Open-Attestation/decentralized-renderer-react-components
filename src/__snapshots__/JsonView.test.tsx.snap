// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonView component should display json view of a json object with nested elements 1`] = `
<pre
  data-testid="json-view"
  style="background-color: white; font-size: 14px; line-height: 21px; white-space: pre-wrap; word-wrap: break-word;"
>
  <div
    style="margin-left: 0px;"
  >
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        string
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        test
      </span>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        number
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        1
      </span>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        boolean
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        false
      </span>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        person
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        <div
          style="margin-left: 40px;"
        >
          <div>
            <span
              style="color: black; font-weight: 700;"
            >
              name
              : 
            </span>
            <span
              style="color: rgb(5, 122, 85);"
            >
              john doe
            </span>
          </div>
          <div>
            <span
              style="color: black; font-weight: 700;"
            >
              age
              : 
            </span>
            <span
              style="color: rgb(5, 122, 85);"
            >
              35
            </span>
          </div>
          <div>
            <span
              style="color: black; font-weight: 700;"
            >
              isMarried
              : 
            </span>
            <span
              style="color: rgb(5, 122, 85);"
            >
              false
            </span>
          </div>
        </div>
      </span>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        nestedObj
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        <div
          style="margin-left: 40px;"
        >
          <div>
            <span
              style="color: black; font-weight: 700;"
            >
              anotherLayerOfNesting
              : 
            </span>
            <span
              style="color: rgb(5, 122, 85);"
            >
              <div
                style="margin-left: 40px;"
              >
                <div>
                  <span
                    style="color: black; font-weight: 700;"
                  >
                    person2
                    : 
                  </span>
                  <span
                    style="color: rgb(5, 122, 85);"
                  >
                    <div
                      style="margin-left: 40px;"
                    >
                      <div>
                        <span
                          style="color: black; font-weight: 700;"
                        >
                          name
                          : 
                        </span>
                        <span
                          style="color: rgb(5, 122, 85);"
                        >
                          jane doe
                        </span>
                      </div>
                      <div>
                        <span
                          style="color: black; font-weight: 700;"
                        >
                          age
                          : 
                        </span>
                        <span
                          style="color: rgb(5, 122, 85);"
                        >
                          20
                        </span>
                      </div>
                      <div>
                        <span
                          style="color: black; font-weight: 700;"
                        >
                          isMarried
                          : 
                        </span>
                        <span
                          style="color: rgb(5, 122, 85);"
                        >
                          true
                        </span>
                      </div>
                    </div>
                  </span>
                </div>
              </div>
            </span>
          </div>
        </div>
      </span>
    </div>
    <div
      style="margin-left: 0px;"
    >
      <div>
        <span
          style="color: black; font-weight: 700;"
        >
          arrayOfObj[0]
          : 
        </span>
        <span
          style="color: rgb(5, 122, 85);"
        >
          <div
            style="margin-left: 40px;"
          >
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                three
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                three
              </span>
            </div>
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                four
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                four
              </span>
            </div>
          </div>
        </span>
      </div>
      <div>
        <span
          style="color: black; font-weight: 700;"
        >
          arrayOfObj[1]
          : 
        </span>
        <span
          style="color: rgb(5, 122, 85);"
        >
          <div
            style="margin-left: 40px;"
          >
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                five
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                five
              </span>
            </div>
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                six
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                six
              </span>
            </div>
          </div>
        </span>
      </div>
    </div>
    <div
      style="margin-left: 0px;"
    >
      <div>
        <span
          style="color: black; font-weight: 700;"
        >
          arrayOfArr[0]
          : 
        </span>
        <span
          style="color: rgb(5, 122, 85);"
        >
          <div
            style="margin-left: 40px;"
          >
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                [0]
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                test3
              </span>
            </div>
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                [1]
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                test4
              </span>
            </div>
          </div>
        </span>
      </div>
      <div>
        <span
          style="color: black; font-weight: 700;"
        >
          arrayOfArr[1]
          : 
        </span>
        <span
          style="color: rgb(5, 122, 85);"
        >
          <div
            style="margin-left: 40px;"
          >
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                [0]
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                test5
              </span>
            </div>
            <div>
              <span
                style="color: black; font-weight: 700;"
              >
                [1]
                : 
              </span>
              <span
                style="color: rgb(5, 122, 85);"
              >
                test6
              </span>
            </div>
          </div>
        </span>
      </div>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        aVeryLongString
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore [...] ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
      </span>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        aVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongKeyName
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore [...] ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
      </span>
    </div>
    <div>
      <span
        style="color: black; font-weight: 700;"
      >
        aVeryNestedObj
        : 
      </span>
      <span
        style="color: rgb(5, 122, 85);"
      >
        <div
          style="margin-left: 40px;"
        >
          <div>
            <span
              style="color: black; font-weight: 700;"
            >
              nest1
              : 
            </span>
            <span
              style="color: rgb(5, 122, 85);"
            >
              <div
                style="margin-left: 40px;"
              >
                <div>
                  <span
                    style="color: black; font-weight: 700;"
                  >
                    nest2
                    : 
                  </span>
                  <span
                    style="color: rgb(5, 122, 85);"
                  >
                    <div
                      style="margin-left: 40px;"
                    >
                      <div>
                        <span
                          style="color: black; font-weight: 700;"
                        >
                          nest3
                          : 
                        </span>
                        <span
                          style="color: rgb(5, 122, 85);"
                        >
                          <div
                            style="margin-left: 40px;"
                          >
                            <div>
                              <span
                                style="color: black; font-weight: 700;"
                              >
                                nest4
                                : 
                              </span>
                              <span
                                style="color: rgb(5, 122, 85);"
                              >
                                <div
                                  style="margin-left: 40px;"
                                >
                                  <div>
                                    <span
                                      style="color: black; font-weight: 700;"
                                    >
                                      nest5
                                      : 
                                    </span>
                                    <span
                                      style="color: rgb(5, 122, 85);"
                                    >
                                      <div
                                        style="margin-left: 40px;"
                                      >
                                        <div>
                                          <span
                                            style="color: black; font-weight: 700;"
                                          >
                                            nest6
                                            : 
                                          </span>
                                          <span
                                            style="color: rgb(5, 122, 85);"
                                          >
                                            <div
                                              style="margin-left: 40px;"
                                            >
                                              <div>
                                                <span
                                                  style="color: black; font-weight: 700;"
                                                >
                                                  nest7
                                                  : 
                                                </span>
                                                <span
                                                  style="color: rgb(5, 122, 85);"
                                                >
                                                  <div
                                                    style="margin-left: 40px;"
                                                  >
                                                    <div>
                                                      <span
                                                        style="color: black; font-weight: 700;"
                                                      >
                                                        nest8
                                                        : 
                                                      </span>
                                                      <span
                                                        style="color: rgb(5, 122, 85);"
                                                      >
                                                        <div
                                                          style="margin-left: 40px;"
                                                        >
                                                          <div>
                                                            <span
                                                              style="color: black; font-weight: 700;"
                                                            >
                                                              nest9
                                                              : 
                                                            </span>
                                                            <span
                                                              style="color: rgb(5, 122, 85);"
                                                            >
                                                              <div
                                                                style="margin-left: 40px;"
                                                              >
                                                                <div>
                                                                  <span
                                                                    style="color: black; font-weight: 700;"
                                                                  >
                                                                    nest10
                                                                    : 
                                                                  </span>
                                                                  <span
                                                                    style="color: rgb(5, 122, 85);"
                                                                  >
                                                                    ...
                                                                  </span>
                                                                </div>
                                                              </div>
                                                            </span>
                                                          </div>
                                                        </div>
                                                      </span>
                                                    </div>
                                                  </div>
                                                </span>
                                              </div>
                                            </div>
                                          </span>
                                        </div>
                                      </div>
                                    </span>
                                  </div>
                                </div>
                              </span>
                            </div>
                          </div>
                        </span>
                      </div>
                    </div>
                  </span>
                </div>
              </div>
            </span>
          </div>
        </div>
      </span>
    </div>
  </div>
</pre>
`;
